### ISSUE TITLE
Mailgun backend loses attachments with non-ASCII filenames

### ISSUE THREAD

**Issue by decibyte** [anymail/django-anymail#125]
* Anymail version: 4.2
* ESP: Mailgun
* Django: 1.11.8
* Python: 3.5

It took me a while to narrow down the circumstances, but it seems that if an attachment has a file name longer than 64 bytes, it's lost. It's simply not part of the e-mail. Neither in my mailbox, nor visible in the Mailgun logs (the attachment list is empty).

It's not just the string length of the file name. If I use special characters (like `æøå`, which we happen to use a lot in Denmark), the allowed string length is shortened.

Initially, I thought this was a Mailgun issue, but I've been in contact with their support, and they claim they don't have this limitation (they _do_ have a limit, but it's higher[1]). Also, if I send an e-mail with an attachment with a very long file name, based on [the `curl` example from their docs](https://documentation.mailgun.com/en/latest/api-sending.html#examples), the attachment is not lost.

If I send the e-mail using Django's built in SMTP backend, everything works fine with very long file names.

On the way, I thought special characters in general was the issue, but that's not the case:

 * The problem persists, even if I encode the file names with [`email.header.make_makeheader`](https://docs.python.org/3/library/email.header.html#email.header.make_header).
 * ASCII-only file names longer than 64 characters also fail.

I've been digging around the source, hoping to find a way to fix it myself, but didn't have any luck. If someone can point me in the right direction, I'd love to provide a fix myself.

[1] Quote from Mailgun support:

> When sending attachments, the size limit is 25mb, so you will also want to verify that your attachment falls under that limit. Regarding the file name length, there can be an argument made for the limit being as high as 255 and as low as 78 depending on the environment. Even if we err on the side of caution and treat 78 as the max allowed, your filename length at 64 should still be acceptable.

**Comment by medmunds** [comment]
That's super weird. If you were hitting some limitation in Anymail or underlying packages (requests, email.mime), I'd expect an exception or outright failure to send, rather than silently dropping the attachment.

Here are some places you might check:

* [`MailgunPayload.add_attachment`](https://github.com/anymail/django-anymail/blob/v4.2/anymail/backends/mailgun.py#L153-L163): is it called? is `attachment.name` what you expect? If not, there may be a bug in the Anymail code that [normalizes attachments](https://github.com/anymail/django-anymail/blob/v4.2/anymail/utils.py#L262-L310).

* [`AnymailRequestsBackend.post_to_esp`](https://github.com/anymail/django-anymail/blob/v4.2/anymail/backends/base_requests.py#L69): this is the handoff to `requests.post`. I wouldn't expect to see anything different here from the previous item, but if `params` (and in particular `params['files']`) doesn't look right, that may provide some other clues. (If you're in a debugger, you might also step into requests to see how it's handling the files param.)

* A definitive test: Capture the raw POST to Mailgun's API by enabling logging on the requests package (see the code block at the end of "API changes" in [these requests docs](http://docs.python-requests.org/en/master/api/#api-changes)). Does the posted body include your attachment with the correct filename? If not, there may be an issue with how Anymail is calling requests, or possibly with the requests package itself (in how it prepares multipart/form-data files). But if your attachment *is* there, the problem is actually at Mailgun's end, and you'll have the exact headers and body to use with curl to demonstrate it to them.

Let me know what you find, and if I can provide any help.

**Comment by costela** [comment]
After debugging a very similar issue for an embarrassing amount of time, I've found something I'm hesitant to call a bug, but which is at least slightly unclear behavior that might be better documented. Maybe there's some overlap with @decibyte's original problem?

It starts with the fact that Mailgun's HTTP API treats downloadable and inline attachments as explicitly distinct parameters. Presumably to enable easier processing of inline images, judging by the `curl` examples in the docs.
In contrast, with SMTP's plain list of MIME parts, the receiving MUA decides what to do with the attachment based on the existence of `Content-Disposition` or `Content-ID` headers. 
The HTTP API design means `django-anymail` has to chose how to send the attachment (whether  inline or not), and this choice is currently done **solely** based on the `Content-Disposition` MIME-header. 
The problem we were seeing popped up when we had inline attachments that were sent *without* a `Content-Disposition` MIME-header, but *with* the `Content-ID` header. This worked as expected with SMTP, because the MUAs we tested were smart enough to infer the attachments were inline. However, in the HTTP API case, `django-anymail` assumed they were "normal" attachments and sent them as such, without filenames and without the `Content-ID` header (which I'm not even sure the HTTP API would accept anyway, since it overloads the `filename` parameter in the `Content-Disposition` header to serve as the `Content-ID` header itself).

And then came the second part of the problem: Mailgun seems to silently drop attachments sent via the HTTP API without filenames.
This gets attached normally:
```
Content-Disposition: form-data; name="attachment"; filename="somefile.png"
Content-Type: image/png
(...)
```
This gets silently dropped:
```
Content-Disposition: form-data; name="attachment"
Content-Type: image/png
(...)
```

Interestingly, trying to send an attachment with `name="inline"` and no filename causes a `500: INTERNAL SERVER ERROR`.

My proposal to make this a bit less confusing is to also check the existence of a `Content-ID` MIME-header and assume - in the absence of an explicit `Content-Disposition: attachment` - that the attachment is inline. This would bring `django-anymail` more in line with the SMTP behavior.


PS.: ideally, I see no reason Mailgun shouldn't also error out (with a 400 code) instead of just silently dropping attachments because it doesn't know what to do with them, but there might be other reasons for this behavior I'm overlooking.

**Comment by medmunds** [comment]
It looks like there may be two separate issues in @decibyte's original report:

1. Definite Anymail+Mailgun issue with sending attachments whose filenames contain *any non-ASCII* characters, no matter the length. I understand what's going wrong, but don't have a solution yet—more below. (And @costela this *is* related to the missing-filename oddity you found in the Mailgun API, but has nothing to do with `Content-ID` or inline attachments.)

2. Possible Mailgun problem with *long* attachment filenames? I'm not able to reproduce this. @decibyte if you have a specific long filename example that fails (and doesn't also involve non-ASCII characters), it'd be really helpful to see.

The first problem turns out to be an interaction between Mailgun's use of `multipart/form-data` for their HTTP API, and the long, confusing history of specs for how to post non-ASCII filenames in HTTP form data.

Here's an example. I'm able to reproduce the first problem like this (using the Mailgun backend):

```python
from django.core.mail import EmailMessage
msg = EmailMessage(to=["medmunds@example.com"], 
                   subject="Mailgun attachment test", 
                   body="This has attachments")
msg.attach("ascii-name.txt", "some data", "text/plain")
msg.attach("unicøde-name.txt", "more data", "text/plain")
msg.send()
```

When this is sent, the attachment with the all-ASCII name makes it through, but the one with a Unicode character is silently dropped.

Here's a snippet of the resulting POST body sent to Mailgun. Check out the filename params:

```http
--35d9a03bfd3f242b57fc32994af09f0c
Content-Disposition: form-data; name="attachment"; filename="ascii-name.txt"
Content-Type: text/plain

some data
--35d9a03bfd3f242b57fc32994af09f0c
Content-Disposition: form-data; name="attachment"; filename*=utf-8''unic%C3%B8de-name.txt
Content-Type: text/plain

more data
--35d9a03bfd3f242b57fc32994af09f0c--
```

That second one, `filename*=utf-8''unic%C3%B8de-name.txt`, uses [RFC 2231](https://tools.ietf.org/html/rfc2231#section-4) character-set encoding. And it's exactly how you'd want to handle that filename in a normal MIME email message body. All the way back to at least 1997.

But the problem is, HTTP multipart/form-data is not exactly MIME. And (historical digression) the specs around how to upload filenames with non-ASCII characters in form-data have been, I guess... slow to emerge? Or at least confusing. Apparently some server implementors went with RFC 2231, and maybe it worked with some clients that had done the same. But more often not. [RFC 5987](https://tools.ietf.org/html/rfc5987) (in 2008) covered using RFC 2231 encoding in HTTP headers, but had an (oft-overlooked?) note specifically excluding RFC 2231 for MIME headers later in a multipart/form-data body.

Jump forward to 2013, and [RFC 7578](https://tools.ietf.org/html/rfc7578#section-4.2) has finally clarified attachment filenames in multipart/form-data: "a 'filename*' parameter [in] the Content-Disposition header field, MUST NOT be used." And instead, "file names normally visible to users MAY be encoded using the percent-encoding method." Also, it notes that in the real world, a bunch of servers will just accept raw UTF-8 bytes as filenames, without offering an opinion on whether that may/should/must be supported.

Anyway, back to Anymail and Mailgun. Here's what's going on: Anymail is trying to send RFC 2231-encoded filenames, and Mailgun doesn't understand that. So much so, that it thinks there simply isn't a filename on that second attachment. And as @costela figured out, Mailgun silently ignores attachments with a missing filename. (Mailgun *doesn't* handle percent-encoded filenames; it sends the attachments, with the %'s verbatim. It *does* accept raw UTF-8 bytes in the filename, which is how curl tests succeed.)

The easy workaround is, don't try to send attachments with non-ASCII filenames.

But what do we do about the bug?

* Fix the underlying problem: don't try to use RFC 2331 encoded filenames in attachment filenames. Anymail uses Requests for all HTTP communication. The Requests issue https://github.com/requests/requests/issues/4652 points out that that package actually delegates to Python's urllib3. So https://github.com/urllib3/urllib3/issues/303 would be where to fix the problem at its root. But even if that issue got addressed soon (which given the history, seems unlikely), it'd take a while to make its way onto production systems.

* Mailgun could issue an API error on missing attachment filenames, rather than silently stripping them. Or better yet, *accept* these attachments, perhaps generating its own filename for them. ([RFC 7578](https://tools.ietf.org/html/rfc7578#section-4.2) weighs in: "The file name isn't mandatory for cases where the file name isn't available or is meaningless or private.") If either of you have an open support case with Mailgun, you might want to suggest this to them.

* Anymail could raise an unsupported feature error in the Mailgun backend on attachments with non-ASCII filenames. (And also missing filenames.) I'm not thrilled with this, but it'd at least let you know something wasn't going to work right.

* Anymail could try to work around the urllib3 issue. I have an idea on this, but it might be too ugly to actually use.

Open to suggestions.

**Comment by costela** [comment]
Thanks @medmunds, that was very informative! I always wondered what was the relation between HTTP's `multipart/form-data` and MIME standards, but was too lazy to go digging in the RFCs.

The PR I proposed should cover our side of the issue, but I'll also open a support ticket with Mailgun, so that at the very least the failure becomes more obvious.
As for fixing the actual issue, I can't think of anything on `django-anymail`'s side short of monkeypatching `urllib3.fields.format_header_param`. I imagine your "ugly" solution went in this direction? Or did you have something else in mind?

**Comment by medmunds** [comment]
Workaround for non-ASCII filenames released in Anymail v4.3.

@decibyte if you are still having issues with *long* filenames, please post an example and reopen. Thanks.
